src/Microsoft.ML.Transforms/Text/TextNormalizerTransform.cs

```diff
@@ -215,171 +156,426 @@ public override void Save(ModelSaveContext ctx)
-            var buffer = new StringBuilder();
-            return
-                (ref ReadOnlyMemory<char> dst) =>
-                {
-                    getSrc(ref src);
-                    NormalizeSrc(ref src, ref dst, buffer);
-                };
+            Contracts.CheckValue(env, nameof(env));
+            env.CheckValue(args, nameof(args));
+            env.CheckValue(input, nameof(input));
+
+            env.CheckValue(args.Column, nameof(args.Column));
+            var cols = new ColumnInfo[args.Column.Length];
+            for (int i = 0; i < cols.Length; i++)
+            {
+                var item = args.Column[i];
+                cols[i] = new ColumnInfo(item.Source ?? item.Name, item.Name,
+                    args.TextCase, args.KeepDiacritics, args.KeepPunctuations, args.KeepNumbers);
+            }
+            return new TextNormalizerTransform(env, cols).MakeDataTransform(input);
         }
 
-        private ValueGetter<VBuffer<ReadOnlyMemory<char>>> MakeGetterVec(IRow input, int iinfo)
+        // Factory method for SignatureLoadDataTransform.
+        private static IDataTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+            => Create(env, ctx).MakeDataTransform(input);
+
+        // Factory method for SignatureLoadRowMapper.
+        private static IRowMapper Create(IHostEnvironment env, ModelLoadContext ctx, ISchema inputSchema)
+            => Create(env, ctx).MakeRowMapper(inputSchema);
+
+        protected override IRowMapper MakeRowMapper(ISchema schema) => new Mapper(this, Schema.Create(schema));
+
+        private sealed class Mapper : MapperBase
         {
-            var getSrc = GetSrcGetter<VBuffer<ReadOnlyMemory<char>>>(input, iinfo);
-            Host.AssertValue(getSrc);
-            var src = default(VBuffer<ReadOnlyMemory<char>>);
-            var buffer = new StringBuilder();
-            var list = new List<ReadOnlyMemory<char>>();
-            var temp = default(ReadOnlyMemory<char>);
-            return
-                (ref VBuffer<ReadOnlyMemory<char>> dst) =>
+            private readonly ColumnType[] _types;
+            private readonly TextNormalizerTransform _parent;
+
+            public Mapper(TextNormalizerTransform parent, Schema inputSchema)
+              : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
+            {
+                _parent = parent;
+                _types = new ColumnType[_parent._columns.Length];
+                for (int i = 0; i < _types.Length; i++)
                 {
-                    getSrc(ref src);
-                    list.Clear();
-                    for (int i = 0; i < src.Count; i++)
+                    inputSchema.TryGetColumnIndex(_parent._columns[i].Input, out int srcCol);
+                    var srcType = inputSchema.GetColumnType(srcCol);
+                    _types[i] = srcType.IsVector ? new VectorType(TextType.Instance) : srcType;
```

Comment:
>VectorType(TextType.Instance) : srcType [](start = 55, length = 39)

I believe you should validate `srcType` here.
Also, are you really creating a variable vector of text? Why do you even change type at all?

Reply:
Oh, I see, it's dropping the empty ones. Wow

---
In reply to: [226072116](https://github.com/dotnet/machinelearning/pull/1276#discussion_r226072116) [](ancestors = 226072116)

Reply:
The validation comment still stands

---
In reply to: [226072884](https://github.com/dotnet/machinelearning/pull/1276#discussion_r226072884) [](ancestors = 226072884,226072116)

Reply:
What kind of validation?

---
In reply to: [226072949](https://github.com/dotnet/machinelearning/pull/1276#discussion_r226072949) [](ancestors = 226072949,226072884,226072116)